tutorials/math/combinatorics.md

```diff
@@ -1,10 +1,215 @@
+- Number of ways to make balanced bracket sequences using $n$ left and $n$ right brackets.
+- [Number of ways to make binary trees](https://leetcode.com/problems/unique-binary-search-trees/description/)
+- Number of ways to form a mountain range with $n$ upstrokes and downstrokes. $\\$
+
+[Here](https://en.wikipedia.org/wiki/Catalan_number#Applications_in_combinatorics) is a more exhastive list.
+
+The $n^{th}$ Catalan number can be found using the formula:
+$C_n$ $=$ $\frac{1}{n + 1}$$2n \choose n$
+
+## Few more examples
+
+### Example #2 [1863 - Sum of All Subset XOR Totals](https://leetcode.com/problems/sum-of-all-subset-xor-totals/description/)
+
+This is an example of a very tricky problem which heavily simplifies after using some Combinatorics and [Bit Manipulation](../../tutorials/math/bit-manipulation.md)
+
+Here we will consider the $i^{th}$ bit from the right. Let's say that the $i^{th}$ bit is set in $k$ out of $n$ numbers in some given subset. If $k$ is odd, then $i^{th}$ bit is set in the XOR of all numbers of the subset, otherwise, it is not set.
+
+Hence if there are $m$ numbers out of $n$ with $i^{th}$ bit set, then the contribution of the bit is $Place\,value\,of\,the\,bit$ $*$ $number\,of\,ways\,to\,get\,odd\,k$ 
+
+Thus we can find $\sum_{k = 1}^{k <= m}$ $m \choose k$ for all odd values of $k$, which comes out to $2^{m - 1}$. Furthermore, we can choose the remaining elements in the subset in $2^{n - m}$ ways by similar logic. Hence total ways to get odd values of $k$ are $2^{n - 1}$, which is independent of both $m$ and $k$.
+
+Hence all we need to do is find bits which are set atleast once (by computing OR) and then multiply the final answer with $2^{n - 1}$.
+Time Complexity of the program is $O(n)$ with $O(1)$ Space Complexity.
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+
+```cpp
+class Solution {
+public:
+    int subsetXORSum(vector<int>& nums) {
+        int arrayOR = 0;
+        // Do OR of whole array to obtain bits which are set atleast once
+        for(int num:nums) arrayOR |= num;
+        // Compute the final answer using the formula discussed
+        return arrayOR*(1<<(nums.size() - 1));
+    }
+};
+```
+
+</TabItem>
+</Tabs>
+
+### Example #3 [62 - Unique Paths](https://leetcode.com/problems/unique-paths/)
+
+Here our robot always goes either down or left. We know that we have to go down $m - 1$ times and go left $n - 1$ times. Thus we need to find the number of ways to arrange these. One way to visualize this is if we have $m + n - 2$ blank spaces, and we have to fill $n - 1$ of them using $L$ and remaining using $D$. Then we can just choose the number of spaces to fill with $L$ from total number of spaces. The the final solution is simply $m + n - 2 \choose n - 1$.
+
+To implement the above, use the template provided above, along with the following code:
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+
+```cpp
+// Uppper limit for m + n - 2 is 198
```

Comment:
typo

besides, please show the full code. verify it in LC first. it gives WA.

Reply:
It gives WA because my code takes mod 1e9 + 7 by default and here we have to return the value as it is. I should have mentioned this though... will add it.